src-electron/rest/user-data.js

```diff
@@ -84,12 +84,123 @@ function httpGetEndpointIds(db) {
  */
 function httpGetDeviceTypeFeatures(db) {
   return async (request, response) => {
-    let deviceTypeRefs = request.query.deviceTypeRefs
-    let deviceTypeFeatures = await queryFeature.getFeaturesByDeviceTypeRefs(
+    let { deviceTypeRefs, endpointTypeRef } = request.query
+    if (Array.isArray(deviceTypeRefs) && deviceTypeRefs.length > 0) {
+      let deviceTypeFeatures = await queryFeature.getFeaturesByDeviceTypeRefs(
+        db,
+        deviceTypeRefs,
+        endpointTypeRef
+      )
+      response.status(StatusCodes.OK).json(deviceTypeFeatures)
+    } else {
+      response.status(StatusCodes.OK).json([])
+    }
+  }
+}
+
+/**
+ * Get all attributes, commands and events in an endpoint type cluster.
+ * @param {*} db
+ * @param {*} endpointTypeClusterId
+ * @param {*} deviceTypeClusterId
+ * @returns elements object containing all attributes, commands and events
+ * in an endpoint type cluster
+ */
+async function getEndpointTypeElements(
+  db,
+  endpointTypeClusterId,
+  deviceTypeClusterId
+) {
+  let elements = {}
+  elements.attributes =
+    await queryAttribute.selectAttributesByEndpointTypeClusterId(
```

Comment:
If you have an async function that calls 3 async functions that are unrelated, a much better pattern is to do Promise.all() on them, so you allow them to execute in parallel.

So instead of:
   await promise1
   await promise2
   await promise3
   
You do:
    Promise.all([promise1, promise2, promise3])

Reply:
Fixed in the new commit